src/mpid/ch4/src/ch4_recv.h

```diff
@@ -136,19 +139,32 @@ MPL_STATIC_INLINE_PREFIX int MPID_Recv_init(void *buf,
             MPIR_ERR_POP(mpi_errno);
         }
 
-        mpi_errno =
-            MPIDI_NM_mpi_recv_init(buf, count, datatype, rank, tag, comm, context_offset, av,
-                                   &(MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(*request)));
+        MPIDI_CH4I_REQUEST(*request, is_local) = 1;
 
-        if (mpi_errno != MPI_SUCCESS) {
-            MPIR_ERR_POP(mpi_errno);
-        }
+        if (!MPIR_Request_is_complete(*request) && !MPIDI_CH4U_REQUEST_IN_PROGRESS(*request)) {
```

Comment:
From #3339 by @hajimefu:

> This might not be quite correct from the patch ordering point of view -- `MPIDI_CH4U_REQUEST_IN_PROGRESS` assumes that the shmmod uses the active message code but it does not at this moment.

I'm not sure this is true. It assumes that the _request_ is using, but I don't think this patch has any shared memory requirements. It won't do much at this point because shared memory don't yet use active messages. But it if did, this code would apply.

Reply:
Let me rephrase my point. `MPIDI_CH4U_REQUEST_IN_PROGRESS` checks `(r)->dev.ch4.am.req->status`. What if shmmod has no idea about the active message (i.e. `(r)->dev.ch4.am.req`), which is true in the current `master`.

Note that `(r)->dev.ch4.am` and `(r)->dev.ch4.shm` are under the same union and you cannot use both at the same time.

Reply:
> Note that (r)->dev.ch4.am and (r)->dev.ch4.shm are under the same union and you cannot use both at the same time.

Hmm. This is a good point. I'll need to look at that.

Reply:
After merging the entire PR this might not be a huge deal, but I noticed this because I was cherry picking this particular commit for my experiment.

Reply:
Yeah, at this point I'm convinced that the ordering here should change, but I think the union thing you mentioned above is actually a problem that needs to be resolved.

Reply:
I've reordered the patches to avoid the patch ordering problem and moved the fields in the `MPIDI_POSIX_request_t` struct to `MPIDI_POSIX_am_request_t` so they won't have the union issues.